### PR TITLE
[FEAT] Support for table extraction in Prompt studio

### PR DESCRIPTION
**Pull request by harini-venkataraman**
## What
PR add the following inclusions
1. Plugin loader for Modifier plugin.
2. Payload changes in prompt studio and prompt studio registry.
3. Input path addition in Structure tool
...

## Why
This enables support for Table enforce type in Prompt Studio
...

## How
Invoking the table extractor in Prompt service.
...

## Can this PR break any existing features? If yes, please list possible items. If no, please explain why. (PS: Admins do not merge the PR without this section filled)
No, this adds a new support in Prompt Studio. 
...

## Relevant Docs
Not applicable
...

## Related Issues or PRs
https://github.com/Zipstack/unstract-sdk/pull/86
https://github.com/Zipstack/unstract-cloud/pull/192

...

## Dependencies Versions / Env Variables
Not applicable

...

## Notes on Testing
Copy the plugins to the respective folder and test the happy path in prompt studio with enforce type - table.
...

## Screenshots
Refer related cloud PRs for Screenshots.( Not added here due to compliance and security reasons )

...

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: backend/backend/urls.py [hunk 1]
```diff
@@ -117,10 +117,17 @@
 # Clone urls
 try:
 
-    import pluggable_apps.clone.urls  # noqa # pylint: disable=unused-import
```
Comment (chandrasekharan-zipstack) [inline backend/backend/urls.py:120]: @harini-venkataraman why was this removed? Will it cause any issues while resolving the URL patterns?

File: backend/backend/urls.py [hunk 2]
```diff
@@ -118,10 +118,17 @@
...
     pass
+
+try:
+    import pluggable_apps.apps.table_settings  # noqa # pylint: disable=unused-import
+
+    urlpatterns += [
+        path("", include("pluggable_apps.apps.table_settings.urls")),
+    ]
+except ImportError:
+    pass
```
Comment (gaya3-zipstack) [inline backend/backend/urls.py:134]: @harini-venkataramanShould we add an info log here to understand if there was any prob in loading libs for easy debugging. IN the past, i have heard that there were some hiccups in understanding the code flow here from @tahierhussain too...

File: backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py [hunk 3]
```diff
@@ -749,6 +772,36 @@ def _fetch_response(
...
         return output_response
 
+    @staticmethod
+    def fetch_table_settings_if_enabled(
+        doc_name: str,
+        prompt: ToolStudioPrompt,
+        org_id: str,
+        user_id: str,
+        tool_id: str,
+        output: dict[str, Any],
```
[the remark this answers is not in the document]
Comment (harini-venkataraman) [reply backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py:782]: @chandrasekharan-zipstack We are using it to pass it to the plugin and update the value of `output`.

File: backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py [hunk 4]
```diff
@@ -426,6 +430,10 @@ def _execute_single_prompt(
         profile_manager_id,
     ):
         prompt_instance = PromptStudioHelper._fetch_prompt_from_id(id)
+
+        if prompt_instance.enforce_type == TSPKeys.TABLE and not modifier_loader:
+            raise OperationNotSupported()
```
[the remark this answers is not in the document]
Comment (harini-venkataraman) [reply backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py:435]: @chandrasekharan-zipstack We have added another check of Single pass, excluding the table enforced prompts being sent to single pass extraction.
`modifier` plugin is a generic plugin pipeline we can have for modifying/updating payloads. It is more specific to table as of now. If any new plugin is added, we can enrich the check, making it specific to the child plugin.

File: backend/plugins/modifier/payload_modifier/payload_modifier.py
[the remark this answers is not in the document]
Comment (harini-venkataraman) [reply backend/plugins/modifier/payload_modifier/payload_modifier.py:1]: I missed it to remove from getting staged. It was written for cloud. Removed.

File: backend/prompt_studio/prompt_studio_registry/prompt_studio_registry_helper.py [hunk 5]
```diff
@@ -317,6 +319,19 @@ def frame_export_json(
             output[JsonSchemaKey.SECTION] = prompt.profile_manager.section
             output[JsonSchemaKey.REINDEX] = prompt.profile_manager.reindex
             output[JsonSchemaKey.EMBEDDING_SUFFIX] = embedding_suffix
+
+            if prompt.enforce_type == "table":
```
Comment (gaya3-zipstack) [inline backend/prompt_studio/prompt_studio_registry/prompt_studio_registry_helper.py:324]: We can reuse PSKeys.TABLE instead of the string @harini-venkataraman

**Review by gaya3-zipstack — APPROVED** [review]
Looks good to me.

File: backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chandrasekharan-zipstack) [reply backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py:436]: NIT: @harini-venkataraman my suggestion was towards naming the variable `modifier_plugins` instead to allow for `for plugin in modifier_plugins:`

File: tools/structure/src/main.py [hunk 6]
```diff
@@ -156,6 +155,18 @@ def run(
...
+            for output in outputs:
+                try:
+                    table_settings = output[SettingsKeys.TABLE_SETTINGS]
+                    extracted_input_file = tool_data_dir / SettingsKeys.EXTRACT
+                    table_settings[SettingsKeys.INPUT_FILE] = extracted_input_file
+                    output.update({SettingsKeys.TABLE_SETTINGS: table_settings})
+
+                except KeyError:
+                    # To check if the prompt has table enforce type selected.
+                    pass
```
Comment (chandrasekharan-zipstack) [inline tools/structure/src/main.py:168]: @Deepak-Kesavan @harini-venkataraman can this be pluggable too?

File: backend/prompt_studio/prompt_studio/migrations/0007_alter_toolstudioprompt_enforce_type.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+                choices=[
+                    ("Text", "Response sent as Text"),
+                    ("number", "Response sent as number"),
+                    ("email", "Response sent as email"),
+                    ("date", "Response sent as date"),
+                    ("boolean", "Response sent as boolean"),
+                    ("json", "Response sent as json"),
+                    ("table", "Response sent as table"),
+                ],
+                db_comment="Field to store the type in             which the response to be returned.",
```
Comment (Deepak-Kesavan) [inline backend/prompt_studio/prompt_studio/migrations/0007_alter_toolstudioprompt_enforce_type.py:27]: @harini-venkataraman Migration file was not updated after changes in model is made.

File: tools/structure/src/main.py (hunk 6 above)
Comment (Deepak-Kesavan) [reply tools/structure/src/main.py:168]: @chandrasekharan-zipstack We can do it. But I feel that it won't be an issue even if it is left like this. We won't be processing it if table extraction plugin is not there in prompt-service

File: prompt-service/src/unstract/prompt_service/main.py [hunk 8]
```diff
@@ -213,6 +213,51 @@ def prompt_processor() -> Any:
                 "Unable to obtain LLM / embedding / vectorDB",
             )
             return APIError(message=msg)
+
+        if output[PSKeys.TYPE] == PSKeys.TABLE:
```
[the remark this answers is not in the document]
Comment (harini-venkataraman) [reply prompt-service/src/unstract/prompt_service/main.py:218]: Addressed

File: tools/structure/src/main.py (hunk 6 above)
Comment (harini-venkataraman) [reply tools/structure/src/main.py:168]: @chandrasekharan-zipstack We dont expose much logic here, and didn't want to make it pluggable since it is in tools. It will not be executed if the table is not chosen.

File: backend/prompt_studio/prompt_studio/migrations/0007_alter_toolstudioprompt_enforce_type.py (hunk 7 above)
Comment (harini-venkataraman) [reply backend/prompt_studio/prompt_studio/migrations/0007_alter_toolstudioprompt_enforce_type.py:27]: @Deepak-Kesavan Are you refering to this [file](https://github.com/Zipstack/unstract/pull/564/files/c561db60ce66db839e236956a584ab7ac29b5c09#diff-289bb837434ab5d22428caf83826719b37d042825a5a9cfe6a34d10581b77468)?
This is up to date with the model.

File: backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (harini-venkataraman) [reply backend/prompt_studio/prompt_studio_core/prompt_studio_helper.py:436]: @chandrasekharan-zipstack We are treating the plugins as plugin pipeline as per the architecture now. Prefer to retain it this way. Will refactor it in any future use cases.

File: tools/structure/src/main.py (hunk 6 above)
Comment (chandrasekharan-zipstack) [reply tools/structure/src/main.py:177]: My suggestion would be to add a TODO comment and create a ticket to mark it as a tech debt. We aren't tracking any such code that go in under this pretext. I feel that this is how cloud related logic creeps into the OSS code.
Even if something doesn't impact the OSS logic directly, it does make us question what a code snippet does and we'll not have a valid answer for that.
Feel free to resolve this conversation

Comment (harini-venkataraman) [reply tools/structure/src/main.py:177]: @chandrasekharan-zipstack Thanks for the suggestion. Will track this under TODO for pluggability in tools.

**Review by chandrasekharan-zipstack — APPROVED** [review]

File: backend/prompt_studio/prompt_studio/migrations/0007_alter_toolstudioprompt_enforce_type.py (hunk 7 above)
Comment (Deepak-Kesavan) [reply backend/prompt_studio/prompt_studio/migrations/0007_alter_toolstudioprompt_enforce_type.py:27]: Got it. Looks like additional spaces in the db_comment is included in the model file itself.

**Review by Deepak-Kesavan — APPROVED** [review]
